### ISSUE TITLE
Add support for en dash

### ISSUE THREAD

**Issue by dmalan** [asciidoctor/asciidoctor#1174]
Per http://asciidoctor.org/docs/asciidoc-writers-guide/, I see that asciidoctor replaces `--` with an em dash (—). Has similar support for en dashes (–) been considered, perhaps whereby `foo--bar` would continue to yield an em dash (wherein `foo` and `bar` are non-numeric strings) but `#--#` would yield an en dash (wherein `#` and `#` are numbers, as would be the case for ranges of dates or page numbers)?

To date, we've manually embedded `&#8211;` where an en dash is called for, though that approach is no longer ideal, now that we're also rendering our AsciiDoc to PDF.

**Comment by skycaptain** [comment]
Definitely +1. Maybe consider the replacment of the substitution by a user-definable hash. Or at least consider a hash depending on the `:lang` attribute.

@jirutka: Indeed, one can easily type an en-dash or em-dash on osx, as one can on windows with alt+`code`-syntax. The thing is, that most monospaced fonts do not distinguish between different dashes, i.e. an en-dash look the same as an em-dash. Coming from source code one can not easily predict the output.

Btw. I have never seen an Em-dash in german literature and the usage is very uncommon [1]. So for other languages the LaTeX-notation might fit perfectly. Finally, using the LaTeX-notation might ease the job for tools like pandoc.

[1] https://de.wikipedia.org/wiki/Geviertstrich#Verwendung

**Comment by farleylai** [comment]
I am using version 1.5.2. When the dash occurs in a link, the auto replacement causes wrong link address in the rendered html. The replacement may be useful in some cases but should be selective or configurable.

**Comment by elextr** [comment]
@farleylai  You can prevent substitution by escaping the -- with a backslash.

**Comment by farleylai** [comment]
@elextr  I knew the workaround but it is counterintuitive since asciidoctor should know the dash is used in a link string.

**Comment by elextr** [comment]
But Asciidoc(tor) doesn't know if the user meant the -- to be substituted.

If substitutions are made contextual it becomes confusing for users, having to remember where it is substituted, and even more so it it is only applied to some substitutions, and then more syntax is needed to force substitutions if the user meant it.

**Comment by farleylai** [comment]
It is straightforward in the case of link addresses.

**Comment by mojavelinux** [comment]
No substitutions should occur in a link (although they currently do). This will be resolved (along with many other link parsing issues) once we switch to a proper inline parser. Right now, links are subject to normal substitutions, which is something we want to get away from.

**Comment by oncletom** [comment]
Oh interesting conversation: I was looking at microtypography options in Asciidoctor. In French, formatting is indeed different:

- `\s*:\s*` should be replaced as `&nbsp;:&nbsp;` — the same for `!`, `?` and `;`
- `"..."` should be replaced as `“...”`

Is there any way to achieve this w/ asciidoctor?

**Comment by mojavelinux** [comment]
Not currently (unless you are willing to implement a Postprocessor extension). This is something we'll need to make configurable...likely based on the lang attribute (which is currently informational only).

**Comment by elextr** [comment]
It would probably be better to have a configuration based on something explicit for this feature. Using lang for things causes problems.  Is French the same in Paris, France, Canada, French Pacific etc? Certainly English isn't the same in the UK, the US, Canada, Australia, New Zealand etc.

Defining an specific attribute to select the version of the feature is probably best:

```
:featurename: specific_value
```

**Comment by mojavelinux** [comment]
Yes, very true, looking just a lang alone isn't enough. What we likely need is to use the key that defines the replacement sets. These things are very well known in the typesetting industry (Chicago Manual of Style) so we just need whatever grouping is formalized.

**Comment by oncletom** [comment]
@elextr indeed, that's why there are regional suffix too, like `fr_CA` for canadian French, `fr_BE` for belgian French etc. Although, going this route, there are dozens of local dialects derived of French which probably don't have any language 'flag'.

I suppose a `lang` corresponds to a set of pre-configured processing features.

**Comment by elextr** [comment]
Yes, there should be a default set of features selected by `:lang:`, probably more than just this feature. 

But it must be possible to override them on an individual basis.  It just isn't possible to predict what features a government, legal or academic institution will want to have different from the country/region default.

(I mentioned government, legal and academic since they are the usual culprits when it comes to mandating strange style requirements :)

**Comment by michael-o** [comment]
Using `--` for en dash and `---`for em dash would also work for German, though em dash is virtually not used in German. en dash is used for ranges. It mostly corresponds to all cases depicted in Wikipedia for en dash in English oder Chicago Manual of Style.

**Comment by schauder** [comment]
There seem to be a relevant amount of people tha do use en-dash with spaces. See http://www.punctuationmatters.com/hyphen-dash-n-dash-and-m-dash/

**Comment by benhourigan** [comment]
+1 for using `--` for en dash and `---` for em dash, like LaTeX.

In my work as an editor I also see a lot of people who use British English using spaced en dashes where *Chicago* would use an em dash. Good to just let the user decide which dash they want. The LaTeX convention is already familiar to many.

**Comment by mojavelinux** [comment]
I played around with this for a bit and I determined that we can't simply link this to compat mode alone since there are many documents not written for compat mode that rely on the `--` replacement for emdash. However, what we can do is leverage the document-wide option mechanism in AsciiDoc. I'm thinking of naming the option `emendash-option` (short for em and en dash option). It would look like this:

```
:emendash-option:

When the package arrived---nearly three weeks late---she no longer needed it.

Working 9--5 like a boss.
```

The `--` can only be used when sandwiched between two word characters. If you need a more esoteric case, you can always just use the glyph. Replacements are meant for convenience, not necessary completeness.

The `--` is replaced by `&#8211;&#8203;`.

I'm open to other suggestions about the name.

I'd like to put this out there behind an option flag so we can start testing it and see how it works out. I anticipate this change is something we will submit as being part of the AsciiDoc specification.

**Comment by elextr** [comment]
> I'd like to put this out there behind an option flag so we can start testing it and see how it works out. I anticipate this change is something we will submit as being part of the AsciiDoc specification.

One of the things that the specification process will need to address beyond just the language is how it manages future changes, doing this means Asciidoctor is also experimenting with one of the possible mechanisms, the `<feature>-option` attribute.

As for the change itself, `--` as endash and `---` as emdash makes sense and has prior art in Latex that suggests it works.

**Comment by jirutka** [comment]
My two cents: I changed this for myself in [asciidoctor-html5s](https://github.com/jirutka/asciidoctor-html5s#replacements):
> Asciidoctor replaces -- with em dash (—) and does not provide any replacement for en dash (–). That’s not very convenient, because en dash is more common than em dash; at least in (British) English and Czech (actually we don’t use em dash at all). So this extension also modifies the replacements table: changes -- to en dash and adds --- for em dash.

**Comment by mojavelinux** [comment]
It's pretty clear we're all in agreement of what we want. The tricky part is figuring out how to get there without breaking tons of documents. That's why I think the option flag is ideal for this (which could be preset in environments down the road).

**Comment by mojavelinux** [comment]
> One of the things that the specification process will need to address beyond just the language is how it manages future changes

A versioned specification (where the documents identify their compliant version). DocBook and DITA both do it with great success. It's clearly something we'll want to consider (though it can't be solved here).

**Comment by michael-o** [comment]
@mojavelinux A breaking change can be easy introduced in 2.0. Until then a switch can do it.

**Comment by mojavelinux** [comment]
@michael-o Unfortunately, when it comes to a change in the AsciiDoc syntax, it's not that simple. GitHub uses Asciidoctor to convert AsciiDoc, but does not allow you to specify a version. Whenever they decide to upgrade Asciidoctor, it immediately rolls out to all users, unless we make it opt in.

**Comment by mojavelinux** [comment]
I'm still wrestling with how we're going to deal with existing documents, so this has to wait. But now that we're on semantic versioning, it will go out when it needs to go out and not get held up.

**Comment by dustypomerleau** [comment]
Thanks to the core team for all your hard work on Asciidoctor.

Since 2.0 has come and gone, can anyone clarify whether there is still intent to allow en dashes at some point?

**Comment by mojavelinux** [comment]
This is a core language change. And with the spec process just days away from getting underway, I don't want to make any changes to the language that would step on the toes of that process. We're going to wait and move forward in concert with the language changes proposed there.

**Comment by dustypomerleau** [comment]
Is it necessary for me to propose it somewhere related to the spec? Or is it already roadmapped?

**Comment by mojavelinux** [comment]
I've marked it here has asciidoc-spec. That way we won't forget when it comes time. The Steering Committee is meeting for the first time this week to vote to launch the project, so while it doesn't exist yet, it will soon. The spec project will have an issue tracker, which is where these requests will land.

**Comment by oelrich** [comment]
Just to add some information on en dash use; in Swedish it is used to represent speach or thought (talstreck/tankestreck).
Basically like this:
- What?

**Comment by dustypomerleau** [comment]
> I've marked it here has asciidoc-spec. That way we won't forget when it comes time. The Steering Committee is meeting for the first time this week to vote to launch the project, so while it doesn't exist yet, it will soon. The spec project will have an issue tracker, which is where these requests will land.

Is the spec repo here on GitHub, or is it hosted elsewhere?

**Comment by mojavelinux** [comment]
The spec project is still in the early stages. You can find information about it here: https://gitlab.eclipse.org/eclipse/asciidoc-lang/asciidoc-lang

**Comment by dustypomerleau** [comment]
> The spec project is still in the early stages. You can find information about it here: https://gitlab.eclipse.org/eclipse/asciidoc-lang/asciidoc-lang

This is what I was looking for - Thanks!

**Comment by jhannwong** [comment]
### To LaTeX, or not to LaTeX?

Is there en dash by now? Would using a `latex` flag help with backward compatibility? If so, it would be nice to have `---` as em dash and `--` as en dash. Is that `latex` mode in the roadmap?

### To Unicode, or not to Unicode?

Meantime, I'll just use Unicode for en dash; easy to replace later on. Not that Unicode is bad. FYI, the LaTeX community has long wanted Unicode (and has it now).

Still, we will need that "_textual replacement_". LaTeX community wants Unicode for language support --- characters not in English, say French Chinese Japanese etc, which are easy to input on the appropriate keyboard layouts --- not for typing hard-to-input Unicode. (In fact, shoehorning English alphabets into "_textually representing_" non-English characters made these characters hard to input, appropriate language-specific keyboards notwithstanding.)

**So, we're still gonna need that en dash, em dash convenient input.**

Also indeed, [the input is the intent](https://github.com/asciidoctor/asciidoctor/issues/1174#issuecomment-453078347).

And some FYI and compliments...

### [Textual Replacements](https://asciidoctor.org/docs/asciidoc-writers-guide/#replacements)

Is great for Unicode that is cumbersome to input. (I just learnt that [Mac does](https://apple.stackexchange.com/a/356566) hyphen, en dash, em dash easy; but I'm not sure I want to move away from tried-and-tested LaTeX conventions; I don't discriminate against Windows users.)

But just letting u know, your support of LaTeX math already adds all the symbols I can ever need. (Yeah, math much, love Asciidoc.)

### Typesetting and Styles

Is good enough in Asciidoctor. LaTeX's typesetting automagic rules does make for beautiful typesetting, but I'm a writer, not an artist (yet). I'm sure any unreadable typesetting in Asciidoctor will be easily and quickly fixed.

The huge advantage that Asciidoc has over LaTeX is productivity; LaTeX is "_markup plus plus plus_" (though the masochism can get strangely addictive, if we're not working hectic outside school).

Asciidoctor's _textual replacement_ engine can surely add even more productivity as needed over time.

**Comment by behrangsa** [comment]
Just a ping. What's the status of supporting (or not supporting) a token to represent en-dashes?
